Add colored output for dcmdump

pjsier opened the pull request:
Progress on #71. I wanted to get some feedback on this initial approach before going much further.

I'm using an enum that implements `fmt::Display` with formatting set for each option. Not sure if I've added the right options to every instance of the values, but I think the overall setup should work. Also calling `condition::set_virtual_terminal` for improving Windows 10 compatibility.

Let me know what you think!

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -52,7 +84,17 @@ where
 
+#[cfg(windows)]
+fn os_compatibility() {
+    control::set_virtual_terminal(true);
```
Look out, this method returns a `Result` which should be handled. Try propagating it to this function and using the existing error report function.

pjsier replied:
It looks like the [original method returns `Result<(), ()>`](https://github.com/mackwic/colored/blob/master/src/control.rs#L28), so I wasn't sure how to use the existing `report` function since that expects an error. I added an exit code instead, but let me know if there's a better approach!

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -263,81 +312,93 @@ where
+        (Str(value), _) => {
+            DumpValue::Str(cut_str(&format!("\"{}\"", value), max_characters).to_string())
+        }
+        (Empty, _) => DumpValue::Num("".to_string()),
```
How about a `DumpValue::Nothing` variant?

Enet4 reviewed (changes requested):
All right, thank you for working on this! Here's how it look on my Windows machine.

![image](https://user-images.githubusercontent.com/4738426/96251085-d6a49f00-0fa7-11eb-8369-cd9393108aa2.png)

The colors seem mostly OK, performance is acceptable too. The trailing pad after the attribute keyword is gone now, but maybe we can make a compromise and set it to an average that would cover most name lengths?
I would also suggest trying a lighter blue for binary values, they are a bit hard to read right now.

Colors are disabled if not on a TTY, just as intended.

Other than that, I did a quick glance at the code, and it seems OK other than the issue noted inline. The idea of keeping color information in an enum is also a good way to prevent inconsistencies.

pjsier commented:
Thanks for the feedback! I addressed some of the inline comments and I'll look more into the trailing pad and updating the blue for binary values, agreed that it's a bit hard to read

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -40,6 +42,7 @@ where
     Str(T),
     DateTime(T),
     Invalid(T),
+    Nothing(T),
```
I was thinking more along the lines of this. If there really is no value to dump, we don't need to take a value.

```suggestion
    Nothing,
```

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -54,6 +57,7 @@ where
             DumpValue::Str(v) => v.to_string().yellow(),
             DumpValue::DateTime(v) => v.to_string().green(),
             DumpValue::Invalid(v) => v.to_string().red(),
+            DumpValue::Nothing(v) => v.to_string().dimmed(),
```
Either one would be OK (don't know of a way to color the second option without allocating though).

```suggestion
            DumpValue::Nothing => return Ok(()),
```

```suggestion
            DumpValue::Nothing => return write!("(no value)"),
```

pjsier replied:
Would this work since it's implemented on `&str`? Still trying to get a handle on allocation and strings though

```rust
DumpValue::Nothing => "(no value)".dimmed(),
```

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -85,15 +89,20 @@ where
 fn main() {
-    os_compatibility();
+    os_compatibility().unwrap_or_else(|_| {
+        println!("Error setting OS compatibility");
```
```suggestion
        eprintln!("Error setting OS compatibility for colored output");
```

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -398,7 +407,7 @@ fn value_summary(value: &PrimitiveValue, vr: VR, max_characters: u32) -> DumpVal
             DumpValue::Str(cut_str(&format!("\"{}\"", value), max_characters).to_string())
         }
-        (Empty, _) => DumpValue::Num("".to_string()),
+        (Empty, _) => DumpValue::Nothing("".to_string()),
```
```suggestion
        (Empty, _) => DumpValue::Nothing,
```

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -85,15 +89,20 @@ where
-    os_compatibility();
+    os_compatibility().unwrap_or_else(|_| {
+        println!("Error setting OS compatibility");
+        std::process::exit(ERROR_COMPAT);
```
Instead of exiting, could the application continue running without colored output?

pjsier commented:
I think I implemented the changes requested here! Also fixed the trailing padding after the attribute keyword. Not sure about whether the `DumpValue::Nothing` string implementation is right though

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -199,7 +251,12 @@ where
+            writeln!(
+                to,
+                "{} {}",
+                DumpValue::TagNum("(FFFE, E0DD)"),
```
Tags are displayed without spaces, one might as well fix existing inconsistencies.
```suggestion
                DumpValue::TagNum("(FFFE,E0DD)"),
```

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -199,7 +251,12 @@ where
+                to,
+                "{} {}",
+                DumpValue::TagNum("(FFFE, E0DD)"),
+                DumpValue::Alias("ItemDelimitationItem"),
```
```suggestion
                DumpValue::Alias("SequenceDelimitationItem"),
```

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -263,81 +322,93 @@ where
 {
     let indent: String = std::iter::repeat(' ').take((depth * 2) as usize).collect();
-    writeln!(to, "{}(FFFE,E000) na Item", indent)?;
+    writeln!(
```
The indentation should go before the tag. Things look weird when dumping sequence items.

```none
(0018,6011) SequenceOfUltrasoundRegions         SQ (2 Items)
(FFFE,E000)     na Item
      (0018,6012) RegionSpatialFormat                 US (1,  2 bytes): 1
      (0018,6014) RegionDataType                      US (1,  2 bytes): 1
      (0018,6016) RegionFlags                         UL (1,  4 bytes): 3
(FFFE,E00D)     ItemDelimitationItem
(FFFE,E000)     na Item
```

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -27,6 +29,42 @@ const ERROR_READ: i32 = -2;
+            DumpValue::Str(v) => v.to_string().yellow(),
+            DumpValue::DateTime(v) => v.to_string().green(),
+            DumpValue::Invalid(v) => v.to_string().red(),
+            DumpValue::Nothing => "(no value)".dimmed(),
```
`dimmed` seems to not make a difference from the rest of the text here. I'd say we either try a darker shade or remove the call to `dimmed()`.

![Capture-dicom-rs-colored-dcmdump-2](https://user-images.githubusercontent.com/4738426/96422094-80c23800-11ef-11eb-9022-eb6f221e3009.PNG)

pjsier replied:
Good to know! I switched this to `italic` since I didn't want to hard-code a darker color in case someone is using a light theme, and it still needs to return a `ColoredString`

Enet4 commented on dcmdump/src/main.rs:
```diff
@@ -27,6 +29,42 @@ const ERROR_READ: i32 = -2;
+
+impl<T> fmt::Display for DumpValue<T>
+where
+    T: ToString,
```
I feel that one could avoid constructing heap allocated strings here, but we can leave that for another time if it requires too many changes.

pjsier replied:
I'm interested in doing this, but I'm not sure how to go about it so it may make sense for a different issue? Not entirely sure how to convert some of the non-string types to `&str` or `String` without allocating in this context, but I'm likely missing something

Enet4 replied:
This could well be a limitation of `colored`. The struct `ColoredString` works with strings internally, whereas a possible implementation of a hypothetical _`ColorisedSomething`_ could work for any type implementing `Display`, without allocating a string. Sounds like an opportunity to take this concern to the respective [repository](https://github.com/mackwic/colored).

Enet4 reviewed (changes requested):
All right, binary data is easier to read now.

I found a few more problems that should not be too hard to fix, then it's good to merge.

Enet4 reviewed (approved):
Thanks again!
